File: src/lib/api/sshServer.ts
import { devApiFetch, hasTauriRuntime, invoke } from "./client";
import type {
  SshConfigImportResult,
  SshServer,
  SshServerConnectionTestInput,
  SshServerTestResult,
  UpsertSshServerInput,
} from "@/types";

export const sshServerApi = {
  list: () =>
    hasTauriRuntime()
      ? invoke<SshServer[]>("list_ssh_servers")
      : devApiFetch<SshServer[]>("/ssh-servers"),
  upsert: (input: UpsertSshServerInput) =>
    hasTauriRuntime()
      ? invoke<SshServer>("upsert_ssh_server", { input })
      : devApiFetch<SshServer>("/ssh-servers", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  delete: (alias: string) =>
    hasTauriRuntime()
      ? invoke<void>("delete_ssh_server", { alias })
      : devApiFetch<void>(`/ssh-servers/${encodeURIComponent(alias)}`, {
          method: "DELETE",
        }),
  importSshConfig: (path?: string | null) =>
    hasTauriRuntime()
      ? invoke<SshConfigImportResult>("import_ssh_config", { path: path ?? null })
      : devApiFetch<SshConfigImportResult>("/ssh-servers/import", {
          method: "POST",
          body: JSON.stringify(path ?? null),
        }),
  test: (alias: string) =>
    hasTauriRuntime()
      ? invoke<SshServerTestResult>("test_ssh_server", { alias })
      : devApiFetch<SshServerTestResult>(
          `/ssh-servers/${encodeURIComponent(alias)}/test`,
          { method: "POST" },
        ),
  testConnection: (input: SshServerConnectionTestInput) =>
    hasTauriRuntime()
      ? invoke<SshServerTestResult>("test_ssh_server_connection", { input })
      : devApiFetch<SshServerTestResult>("/ssh-servers/test-connection", {
          method: "POST",
          body: JSON.stringify(input),
        }),
};

File: src-tauri/src/commands/ssh_server.rs
use crate::error::CommandError;
use crate::models::{
    SshConfigImportResult, SshServer, SshServerConnectionTestInput, SshServerTestResult,
    UpsertSshServerInput,
};
use crate::services::ssh_server::SshServerService;
use crate::state::AppState;

#[tauri::command]
pub fn list_ssh_servers(state: tauri::State<'_, AppState>) -> Result<Vec<SshServer>, CommandError> {
    SshServerService::list(&state.db).map_err(|e| e.into())
}

#[tauri::command]
pub fn upsert_ssh_server(
    state: tauri::State<'_, AppState>,
    input: UpsertSshServerInput,
) -> Result<SshServer, CommandError> {
    SshServerService::upsert(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn delete_ssh_server(
    state: tauri::State<'_, AppState>,
    alias: String,
) -> Result<(), CommandError> {
    SshServerService::delete(&state.db, &alias).map_err(|e| e.into())
}

#[tauri::command]
pub fn import_ssh_config(
    state: tauri::State<'_, AppState>,
    path: Option<String>,
) -> Result<SshConfigImportResult, CommandError> {
    SshServerService::import_ssh_config(&state.db, path).map_err(|e| e.into())
}

#[tauri::command]
pub async fn test_ssh_server(
    state: tauri::State<'_, AppState>,
    alias: String,
) -> Result<SshServerTestResult, CommandError> {
    SshServerService::test(&state.db, &alias)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn test_ssh_server_connection(
    input: SshServerConnectionTestInput,
) -> Result<SshServerTestResult, CommandError> {
    SshServerService::test_connection(input)
        .await
        .map_err(|e| e.into())
}
